### ISSUE TITLE
Setting up the MongoDB Operator to interact with multiple DocumentDB or MongoDB instances?

### ISSUE THREAD

**Issue by GamerM** [SDA-SE/mongodb-operator#551]
Hello,

Thanks for setting up the MongoDB operator. I was wondering how would I set it up if I would need to interact with multiple DocumentDB clusters? The current solution I've setup is to have 2 kustomize projects with different `nameSuffix` applied on it. The kustomize projects will then have different `mongodbConnectionString` to authenticate to the different DocumentDB clusters. Following the [usage kustomization guide](https://sda-se.github.io/mongodb-operator/usage/#kustomize), I noticed after applying it to the cluster it doesn't apply the suffix to the `mongodb` CRD itself. Which causes it to be shared by both projects.

How can I reference the `mongodb` kind to interact with a specific DocumentDB cluster, if so?

## Current working directory
```
├── documentdb1
│   ├── kustomization.yaml
│   ├── mongodb-configuration.yaml
│   └── resources
│       └── global-bundle.pem
└── documentdb2
    ├── kustomization.yaml
    ├── mongodb-configuration.yaml
    └── resources
        └── global-bundle.pem
```

## documentdb1 kustomization.yaml
```
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: custom-namespace
nameSuffix: "-documentdb1"
configurations:
  - mongodb-configuration.yaml
resources:
  - https://github.com/SDA-SE/mongodb-operator//kustomize/release/no-namespace?ref=1.0.39
secretGenerator:
  - name: mongodb-operator
    literals:
      - mongodbConnectionString=mongodb://username:password@documentdb1.com:27017/?tls=true
configMapGenerator:
  - name: mongodb-operator-ca-pem
    behavior: merge
    files:
      - resources/global-bundle.pem
```

## documentdb2 kustomization.yaml
```
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: custom-namespace
nameSuffix: "-documentdb2"
configurations:
  - mongodb-configuration.yaml
resources:
  - https://github.com/SDA-SE/mongodb-operator//kustomize/release/no-namespace?ref=1.0.39
secretGenerator:
  - name: mongodb-operator
    literals:
      - mongodbConnectionString=mongodb://username:password@documentdb2.com:27017/?tls=true
configMapGenerator:
  - name: mongodb-operator-ca-pem
    behavior: merge
    files:
      - resources/global-bundle.pem
```

The [mongodb-configuration.yaml](https://sda-se.github.io/mongodb-operator/usage/#kustomize) exact copy from the documentation

**Comment by JoergSiebahn** [comment]
Hi @GamerM,

sorry for the late response. We have missed your issue.

Currently, connecting to two different MongoDB servers from one cluster is not in scope of the MongoDB Operator.
But it should be possible to deploy two instances of the MongoDB Operator connected to different clusters in separate namespaces.
To avoid that they interfere, they must be configured to watch distinct resources.

That distinction can't be based on the CRD, because it is not configurable to watch a different kind of resource.
Also, the CRD is not a namespaced resource and can be applied only once (or is updated from multiple kustomizations).
In case two Operators are deployed, only one should apply the CRD or it should be applied independently from the Operator deployments.

It has not be nested, but restricting access based on resource names by modifying the [ClusterRole](https://github.com/SDA-SE/mongodb-operator/blob/master/kustomize/bases/operator/mongodb-operator-cr.yaml) for each deployment to define a fixed set of database names for each namespace (the namespace will still be part of the database name).
It could also be an option to remove the [ClusterRoleBinding](https://github.com/SDA-SE/mongodb-operator/blob/master/kustomize/bases/operator/mongodb-operator-crb.yaml) and create a RoleBinding in each namespace that needs a MongoDB from a specific Operator instance.
I have not tried it, so you may need a deep dive into [Kubernetes RBAC authorization](https://kubernetes.io/docs/reference/access-authn-authz/rbac/).

Maybe this example can help you as a starting point. I have not tried it in a cluster, only checked the output of `kustomize build`:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: mongodb-operator-dev
nameSuffix: -dev

resources:
  # Use kustomization with namespace, patch name below
  - https://github.com/SDA-SE/mongodb-operator//kustomize/release?ref=1.0.39

secretGenerator:
  # Please use SealedSecret, ExternalSecret or similar
  - name: mongodb-operator
    literals:
      - mongodbConnectionString=mongodb://username:password@documentdb1.com:27017/?tls=true

patches:
  # Change the namespace resource to match the defined namespace
  - target:
      kind: Namespace
    patch: |
      - op: replace
        path: /metadata/name
        value: mongodb-operator-dev
  # Do not create the custom resource definition here
  - target:
      kind: CustomResourceDefinition
      name: mongodbs.persistence.sda-se.com
    patch: |
      $patch: delete
      apiVersion: apiextensions.k8s.io/v1
      kind: CustomResourceDefinition
      metadata:
        name: ignored
  # Restrict access to dev-db only
  # This has not been tested and there may be way more useful
  # ways to use RBAC authorization to use multiple DB servers.
  - target:
      kind: ClusterRole
      name: mongodb-operator
    patch: |
      - op: add
        path: /rules/1/resourceNames
        value:
          - dev-db
```

In the long term, supporting multiple database instances, effectively multiple MongoDB Operators in one cluster as you tried, would for sure be a nice feature for the MongoDB Operator.

Adding a configuration for the namespaces, a single MongoDB Operator instance watches, would be an option, that we had that in mind a while ago. I don't remember if we found some caveats that stopped us implementing it or if we just did not have enough time.

Another option could be to introduce a label selector configuration like other Kubernetes resources have.

Can you explain your case a little bit more detailed? Especially in which way you need databases on different database servers and which restrictions apply.

Unfortunately, new features for this project have a very low priority on our side at the moment, because „it works for us“. But we are open for contributions. If you are interested, I'm happy to assist and discuss options.
